packages/suite/src/hooks/wallet/form/useUtxoSelection.ts

```diff
@@ -0,0 +1,103 @@
+
+import { UseSendFormState } from '@suite-common/wallet-types';
+import type { AccountUtxo, PROTO } from '@trezor/connect';
+
+type Props = UseFormMethods &
+    Pick<UseSendFormState, 'account' | 'composedLevels' | 'feeInfo'> & {
+        composeRequest: (field?: string) => void;
+    };
+
+export const useUtxoSelection = ({
+    account,
+    composedLevels,
+    composeRequest,
+    feeInfo,
+    register,
+    setValue,
+    watch,
+}: Props) => {
+    // register custom form field (without HTMLElement)
+    useEffect(() => {
+        register({ name: 'isCoinControl', type: 'custom' });
+        register({ name: 'selectedUtxos', type: 'custom' });
+    }, [register]);
+
+    const isCoinControl = watch('isCoinControl');
+    const selectedUtxos: AccountUtxo[] = watch('selectedUtxos') || [];
+    const selectedFee = watch('selectedFee');
+
+    const [spendableUtxos, dustUtxos]: [AccountUtxo[], AccountUtxo[]] = account.utxo
+        ? account.utxo.reduce(
+              ([previousSpendable, previousDust]: [AccountUtxo[], AccountUtxo[]], current) =>
+                  feeInfo.dustLimit && parseInt(current.amount, 10) >= feeInfo.dustLimit
+                      ? [[...previousSpendable, current], previousDust]
+                      : [previousSpendable, [...previousDust, current]],
+              [[], []],
+          )
+        : [[], []];
+
+    const allUtxosSelected =
+        !!selectedUtxos.length && selectedUtxos.length === spendableUtxos.length;
+    const composedLevel = composedLevels?.[selectedFee || 'normal'];
+
+    const composedInputs = useMemo(
+        () =>
+            composedLevel && 'transaction' in composedLevel ? composedLevel.transaction.inputs : [],
+        [composedLevel],
+    ) as PROTO.TxInputType[];
+    const preselectedUtxos = useMemo(
+        () =>
+            account.utxo?.filter(utxo =>
+                composedInputs.some(
+                    input => input.prev_hash === utxo.txid && input.prev_index === utxo.vout,
+                ),
+            ) || [],
+        [account.utxo, composedInputs],
+    );
+
+    const toggleCheckAllUtxos = () => {
+        if (allUtxosSelected) {
+            setValue('selectedUtxos', []);
+        } else {
+            setValue('selectedUtxos', spendableUtxos);
+            setValue('isCoinControl', true);
+        }
+        composeRequest();
+    };
+
+    const toggleCoinControl = () => {
+        setValue('isCoinControl', !isCoinControl);
+        setValue('selectedUtxos', isCoinControl ? [] : preselectedUtxos);
+        composeRequest();
+    };
+
+    const toggleUtxoSelection = (utxo: AccountUtxo) => {
+        const isSelected = selectedUtxos.find(u => u.txid === utxo.txid && u.vout === utxo.vout);
+        if (isSelected) {
+            setValue('selectedUtxos', selectedUtxos.filter(u => u !== isSelected) || []);
+        } else {
+            setValue(
+                'selectedUtxos',
+                !selectedUtxos.length && !isCoinControl
```

Comment:
nit: define that outside of `setValue()` for readability

Reply:
65b620e49e050a6fedd6e47a68d29cb9276f3454